I get a test failure:
```
[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] Running eu.fasten.vulnerabilityproducer.db.NitriteControllerTest
09:44:12,986 |-INFO in ch.qos.logback.classic.LoggerContext[default] - Could NOT find resource [logback-test.xml]
09:44:12,986 |-INFO in ch.qos.logback.classic.LoggerContext[default] - Found resource [logback.xml] at [file:/c:/Users/Magiel%20Bruntink/repos/vulnerability-producer/target/classes/logback.xml]
09:44:13,018 |-INFO in ch.qos.logback.classic.joran.action.ConfigurationAction - debug attribute not set
09:44:13,033 |-INFO in ch.qos.logback.core.joran.action.StatusListenerAction - Added status listener of type [ch.qos.logback.core.status.OnConsoleStatusListener]
09:44:13,033 |-INFO in ch.qos.logback.core.joran.action.TimestampAction - Using current interpretation time, i.e. now, as time reference.
09:44:13,033 |-INFO in ch.qos.logback.core.joran.action.TimestampAction - Adding property to the context with key="byDay" and value="20220321T094413" to the LOCAL scope
09:44:13,033 |-INFO in ch.qos.logback.core.joran.action.AppenderAction - About to instantiate appender of type [ch.qos.logback.core.ConsoleAppender]
09:44:13,033 |-INFO in ch.qos.logback.core.joran.action.AppenderAction - Naming appender as [STDOUT]
09:44:13,033 |-INFO in ch.qos.logback.core.joran.action.NestedComplexPropertyIA - Assuming default type [ch.qos.logback.classic.encoder.PatternLayoutEncoder] for [encoder] property
09:44:13,049 |-INFO in ch.qos.logback.core.joran.action.AppenderAction - About to instantiate appender of type [ch.qos.logback.core.ConsoleAppender]
09:44:13,049 |-INFO in ch.qos.logback.core.joran.action.AppenderAction - Naming appender as [STDOUT-ERROR]
09:44:13,049 |-INFO in ch.qos.logback.core.joran.action.NestedComplexPropertyIA - Assuming default type [ch.qos.logback.classic.encoder.PatternLayoutEncoder] for [encoder] property
09:44:13,049 |-INFO in ch.qos.logback.core.joran.action.AppenderAction - About to instantiate appender of type [ch.qos.logback.core.FileAppender]
09:44:13,049 |-INFO in ch.qos.logback.core.joran.action.AppenderAction - Naming appender as [FILE]
09:44:13,049 |-INFO in ch.qos.logback.core.joran.action.NestedComplexPropertyIA - Assuming default type [ch.qos.logback.classic.encoder.PatternLayoutEncoder] for [encoder] property
09:44:13,049 |-INFO in ch.qos.logback.core.FileAppender[FILE] - File property is set to [/mnt/fasten/vulnerabilities/producer_logs/out-20220321T094413.log]
09:44:13,049 |-INFO in ch.qos.logback.classic.joran.action.RootLoggerAction - Setting level of ROOT logger to DEBUG
09:44:13,049 |-INFO in ch.qos.logback.core.joran.action.AppenderRefAction - Attaching appender named [STDOUT] to Logger[ROOT]
09:44:13,049 |-INFO in ch.qos.logback.core.joran.action.AppenderRefAction - Attaching appender named [STDOUT-ERROR] to Logger[ROOT]
09:44:13,049 |-INFO in ch.qos.logback.core.joran.action.AppenderRefAction - Attaching appender named [FILE] to Logger[ROOT]
09:44:13,049 |-INFO in ch.qos.logback.classic.joran.action.LoggerAction - Setting level of logger [org.eclipse.jgit.internal.storage.file.FileSnapshot] to OFF
09:44:13,049 |-INFO in ch.qos.logback.classic.joran.action.LoggerAction - Setting level of logger [org.eclipse.jgit.transport.PacketLineIn] to OFF
09:44:13,049 |-INFO in ch.qos.logback.classic.joran.action.LoggerAction - Setting level of logger [org.eclipse.jgit.transport.PacketLineOut] to OFF
09:44:13,049 |-INFO in ch.qos.logback.classic.joran.action.LoggerAction - Setting level of logger [org.eclipse.jgit.util.FS] to OFF
09:44:13,049 |-INFO in ch.qos.logback.classic.joran.action.LoggerAction - Setting level of logger [org.mongodb.driver.cluster] to OFF
09:44:13,049 |-INFO in ch.qos.logback.classic.joran.action.LoggerAction - Setting level of logger [org.mongodb.driver.connection] to OFF
09:44:13,049 |-INFO in ch.qos.logback.classic.joran.action.LoggerAction - Setting level of logger [org.mongodb.driver.protocol.command] to OFF
09:44:13,049 |-INFO in ch.qos.logback.classic.joran.action.LoggerAction - Setting level of logger [org.dizitart.no2.internals.DataService] to OFF
09:44:13,049 |-INFO in ch.qos.logback.classic.joran.action.LoggerAction - Setting level of logger [org.apache.kafka.clients.Metadata] to OFF
09:44:13,049 |-INFO in ch.qos.logback.classic.joran.action.LoggerAction - Setting level of logger [org.apache.kafka.clients.NetworkClient] to OFF
09:44:13,049 |-INFO in ch.qos.logback.classic.joran.action.LoggerAction - Setting level of logger [org.apache.kafka.common.metrics.Metrics] to OFF
09:44:13,049 |-INFO in ch.qos.logback.classic.joran.action.LoggerAction - Setting level of logger [org.apache.kafka.common.network.Selector] to OFF
09:44:13,049 |-INFO in ch.qos.logback.classic.joran.action.ConfigurationAction - End of configuration.
09:44:13,049 |-INFO in ch.qos.logback.classic.joran.JoranConfigurator@4b41dd5c - Registering current configuration as safe fallback point
[2022-03-21 09:44:13,205] [DEBUG] [main] [u.c.j.p.a.PodamFactoryImpl] - Manufacturing class java.lang.String with parameters []
[2022-03-21 09:44:13,205] [DEBUG] [main] [u.c.j.p.a.PodamFactoryImpl] - Populating pojo class java.lang.String
[2022-03-21 09:44:13,221] [DEBUG] [main] [u.c.j.p.a.PodamFactoryImpl] - Manufacturing class java.lang.String with parameters []
[2022-03-21 09:44:13,221] [DEBUG] [main] [u.c.j.p.a.PodamFactoryImpl] - Populating pojo class java.lang.String
[2022-03-21 09:44:13,221] [DEBUG] [main] [u.c.j.p.a.PodamFactoryImpl] - Manufacturing class java.lang.String with parameters []
[2022-03-21 09:44:13,236] [DEBUG] [main] [u.c.j.p.a.PodamFactoryImpl] - Populating pojo class java.lang.String
[2022-03-21 09:44:13,236] [DEBUG] [main] [u.c.j.p.a.PodamFactoryImpl] - Manufacturing class java.lang.String with parameters []
[2022-03-21 09:44:13,236] [DEBUG] [main] [u.c.j.p.a.PodamFactoryImpl] - Populating pojo class java.lang.String
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.516 s - in eu.fasten.vulnerabilityproducer.db.NitriteControllerTest
[INFO] Running eu.fasten.vulnerabilityproducer.mappers.PoolTest
[INFO] Tests run: 3, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0 s - in eu.fasten.vulnerabilityproducer.mappers.PoolTest
[INFO] Running eu.fasten.vulnerabilityproducer.mappers.PurlMapperTest
[INFO] Tests run: 13, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.503 s - in eu.fasten.vulnerabilityproducer.mappers.PurlMapperTest
[INFO] Running eu.fasten.vulnerabilityproducer.mappers.VersionRangerTest
[INFO] Tests run: 27, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.235 s - in eu.fasten.vulnerabilityproducer.mappers.VersionRangerTest
[INFO] Running eu.fasten.vulnerabilityproducer.parsers.ExtraParserTest
[2022-03-21 09:44:14,099] [INFO ] [main] [e.f.v.u.p.ExtraParser] - Found MSR2020 CPP dataset from memory
[2022-03-21 09:44:14,099] [INFO ] [main] [e.f.v.u.p.ExtraParser] - Parsed Vulnerability with ID - CVE-2009-1194
[2022-03-21 09:44:14,115] [INFO ] [main] [e.f.v.u.p.ExtraParser] - Found a total of 1 vulnerability in the CPP dataset from MSR2020
[2022-03-21 09:44:14,115] [INFO ] [main] [e.f.v.u.p.ExtraParser] - Parsed Vulnerability with ID - CVE-2017-4971
[2022-03-21 09:44:14,115] [INFO ] [main] [e.f.v.u.p.ExtraParser] - Parsed Vulnerability with ID - CVE-2018-1000134
[2022-03-21 09:44:14,131] [INFO ] [main] [e.f.v.u.p.ExtraParser] - Parsing statements for GEM: curl
[2022-03-21 09:44:14,177] [INFO ] [main] [e.f.v.u.p.ExtraParser] - Parsed Vulnerability with ID - CVE-2013-2617
[2022-03-21 09:44:14,177] [INFO ] [main] [e.f.v.u.p.ExtraParser] - Parsed 1 from ExtraSources
[2022-03-21 09:44:14,177] [INFO ] [main] [e.f.v.u.p.ExtraParser] - Downloading raw CSV file from MSR2020
[2022-03-21 09:44:14,177] [INFO ] [main] [e.f.v.u.p.ExtraParser] - Writing dataset to memory
[2022-03-21 09:44:14,177] [INFO ] [main] [e.f.v.u.p.ExtraParser] - Parsed Vulnerability with ID - CVE-2009-1194
[2022-03-21 09:44:14,177] [INFO ] [main] [e.f.v.u.p.ExtraParser] - Found a total of 1 vulnerability in the CPP dataset from MSR2020
[2022-03-21 09:44:14,193] [INFO ] [main] [e.f.v.u.p.ExtraParser] - Parsed Vulnerability with ID - CVE-TEST-2
[2022-03-21 09:44:14,193] [INFO ] [main] [e.f.v.u.p.ExtraParser] - Parsed Vulnerability with ID - CVE-TEST-3
[2022-03-21 09:44:14,224] [INFO ] [main] [e.f.v.u.p.ExtraParser] - Parsing java statement file 0001.yaml from year 2020
[2022-03-21 09:44:14,224] [INFO ] [main] [e.f.v.u.p.ExtraParser] - Parsed Vulnerability with ID - CVE-2020-0001
[2022-03-21 09:44:14,224] [INFO ] [main] [e.f.v.u.p.ExtraParser] - Parsing python statement file 0002.yaml from year 2020
[2022-03-21 09:44:14,240] [INFO ] [main] [e.f.v.u.p.ExtraParser] - Parsed Vulnerability with ID - CVE-2020-0002
[2022-03-21 09:44:14,240] [INFO ] [main] [e.f.v.u.p.ExtraParser] - Parsing statement file for CVE-2020-11989
[2022-03-21 09:44:14,240] [INFO ] [main] [e.f.v.u.p.ExtraParser] - Parsed Vulnerability with ID - CVE-2020-11989
[ERROR] Tests run: 9, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 0.141 s <<< FAILURE! - in eu.fasten.vulnerabilityproducer.parsers.ExtraParserTest
[ERROR] testInjectFromSafetyDB  Time elapsed: 0.047 s  <<< FAILURE!
org.opentest4j.AssertionFailedError: expected: <Vulnerability{id='mock-id', purls=[pkg:pypi/mock@0.9, pkg:pypi/mock@1.0], first_patched=[], scoreCVSS2=null, scoreCVSS3=null, severity=null, published_date='null', last_modified_date='null', description='mock description', references=[], patch_links=[], exploits=[], patches=[]}> but was: <null>
	at eu.fasten.vulnerabilityproducer.parsers.ExtraParserTest.testInjectFromSafetyDB(ExtraParserTest.java:109)
```

--- a/src/main/java/eu/fasten/vulnerabilityproducer/Main.java
+++ b/src/main/java/eu/fasten/vulnerabilityproducer/Main.java
@@ -70,6 +70,12 @@ public class Main implements Runnable {
             defaultValue = "both")
     String inferStrategy;
 
+    @CommandLine.Option(names = {"-nt", "--num_threads"},
+            paramLabel = "NoOfThreadsToUse",
+            description = "Sets number of threads for gathering and parsing vulnerability data. [default=6]",
+            defaultValue = "6")
+    int noOfThreadsToUse;
+
     public static void main(String[] args) {
         final int exitCode = new CommandLine(new Main()).execute(args);
         System.exit(exitCode);
@@ -87,6 +93,7 @@ public class Main implements Runnable {
         vulnProducer.setServerAddresses(kafkaServers);
         vulnProducer.setKafkaTopic(kafkaTopic);
         vulnProducer.createKafkaProducer();
+        vulnProducer.setNumThreadsToUse(noOfThreadsToUse);
 
         if (jsonPath != null) {
             // Directly produce the vulnerabilities to the kafka topic

--- a/src/main/java/eu/fasten/vulnerabilityproducer/VulnerabilityProducer.java
+++ b/src/main/java/eu/fasten/vulnerabilityproducer/VulnerabilityProducer.java
@@ -49,6 +49,7 @@ public class VulnerabilityProducer {
     private ParserManager parserManager;
     private MongoDatabase mongoDatabase;
     public String pathToMnt;
+    private int maxNumThreadsToUse;
     private final Logger logger = LoggerFactory.getLogger(VulnerabilityProducer.class.getName());
 
     public void setPathToMnt(String mnt) {
@@ -70,6 +71,8 @@ public class VulnerabilityProducer {
     public void setMongoDatabase(MongoDatabase mongoDatabase) {
         this.mongoDatabase = mongoDatabase;
     }
+
+    public void setNumThreadsToUse(int numThreads) { this.maxNumThreadsToUse = numThreads; }
 
     /**
      * Sets the connection to Mongodb instance.
@@ -105,7 +108,7 @@ public class VulnerabilityProducer {
     public void start(boolean updatesOnly, String inferStrategy) throws IOException {
         createParserManager(inferStrategy);
         do {
-            parserManager.getVulnerabilitiesFromParsers(updatesOnly);
+            parserManager.getVulnerabilitiesFromParsers(updatesOnly, maxNumThreadsToUse);
             logger.info("Sleeping a day, see you in a bit");
             parserManager.sleep();
         } while(true);

--- a/src/main/java/eu/fasten/vulnerabilityproducer/utils/parsers/ExtraParser.java
+++ b/src/main/java/eu/fasten/vulnerabilityproducer/utils/parsers/ExtraParser.java
@@ -104,21 +104,21 @@ public class ExtraParser implements VulnerabilityParser {
 
             var jsonObject = new JSONObject(insecureJson.trim());
             Iterator<String> keys = jsonObject.keys();
-            var insecureVulns = new HashMap<String, JSONObject>();
+            var insecureVulns = new HashMap<String, List<JSONObject>>();
 
             while (keys.hasNext()) {
                 var key = keys.next();
                 var listVulns = jsonObject.optJSONArray(key);
                 if(Objects.nonNull(listVulns)) {
+                    insecureVulns.put(key, new ArrayList<>());
                     for (Object listVuln : listVulns) {
                         var vulnObject = (JSONObject) listVuln;
-                        insecureVulns.put(key, vulnObject);
+                        insecureVulns.get(key).add(vulnObject);
                     }
                 }
             }
 
-            insecureVulns.keySet().parallelStream().forEach((v) -> {parseJSONVulnObj(v, insecureVulns.get(v));});
-
+            insecureVulns.keySet().parallelStream().forEach((v) -> {insecureVulns.get(v).forEach((l) -> {parseJSONVulnObj(v, l);});});
         } catch (Exception e) {
             e.printStackTrace();
         }

--- a/src/main/java/eu/fasten/vulnerabilityproducer/utils/parsers/ParserManager.java
+++ b/src/main/java/eu/fasten/vulnerabilityproducer/utils/parsers/ParserManager.java
@@ -98,7 +98,7 @@ public class ParserManager {
         this.kafkaTopic = kafkaTopic;
     }
 
-    public void getVulnerabilitiesFromParsers(boolean updatesOnly) {
+    public void getVulnerabilitiesFromParsers(boolean updatesOnly, int numThreadsToUse) {
         if(updatesOnly) {
             logger.info("Updating Versions of Packages");
             versionRanger.updateMappings();
@@ -112,26 +112,41 @@ public class ParserManager {
         Thread smtWriter = new Thread(() -> smtQueue.forEach(this::storeStatement));
         smtWriter.start();
 
-        logger.info("Injecting information about Patches");
+        logger.info("Injecting information about Patches  using " + numThreadsToUse  + " threads");
         var vulnerabilitiesWithForge = vulnerabilities.stream().filter(this::vulnerabilityForgeIsSupported).collect(Collectors.toList());
-        ProgressBar.wrap(vulnerabilitiesWithForge.parallelStream(), "InjectPatches").forEach((v) -> {
-            try {
-                var patchFinder = patchFinderFactory.getNewPatchFinder();
-                logger.info("Inferring purls for " + v.getId());
-                purlMapper.inferPurls(v, patchFinder);
-                logger.info("Looking for patches in references of " + v.getId());
-                patchFinder.parseReferences(v, nitriteController);
-                logger.info("Publishing " + v.getId() + " to Kafka");
-                emitKafkaMsg(v.toJson());
-                smtQueue.add(v);
-            } catch (Exception e) {
-                logger.error("Could NOT process " + v.getId() + "\n" + e);
+
+        var executor = Executors.newFixedThreadPool(numThreadsToUse);
+        List<Callable<Object>> patchFindingTasks = new ArrayList<>();
+        try(ProgressBar pb = new ProgressBar("InjectPatches", vulnerabilitiesWithForge.size())) {
+            for (var v: vulnerabilitiesWithForge) {
+                patchFindingTasks.add(() -> {parsePatchesAndPublish(smtQueue, v, pb);
+                    return null;
+                });
             }
-        });
+            try {
+                executor.invokeAll(patchFindingTasks);
+                executor.shutdown();
+                smtWriter.join();
+            } catch (InterruptedException e) {
+                throw new RuntimeException(e);
+            }
+        }
+    }
+
+    private void parsePatchesAndPublish(ConcurrentLinkedQueue<Vulnerability> smtQueue, Vulnerability v, ProgressBar pb) {
         try {
-            smtWriter.join();
-        } catch (InterruptedException e) {
-            throw new RuntimeException(e);
+            var patchFinder = patchFinderFactory.getNewPatchFinder();
+            logger.info("Inferring purls for " + v.getId());
+            purlMapper.inferPurls(v, patchFinder);
+            logger.info("Looking for patches in references of " + v.getId());
+            patchFinder.parseReferences(v, nitriteController);
+            logger.info("Publishing " + v.getId() + " to Kafka");
+            emitKafkaMsg(v.toJson());
+            smtQueue.add(v);
+            pb.step();
+        } catch (Exception e) {
+            logger.error("Could NOT process " + v.getId() + "\n" + e);
+            pb.step();
         }
     }
 

--- a/src/test/java/eu/fasten/vulnerabilityproducer/parsers/ParserManagerTest.java
+++ b/src/test/java/eu/fasten/vulnerabilityproducer/parsers/ParserManagerTest.java
@@ -128,7 +128,7 @@ public class ParserManagerTest {
         when(patchFinderFactoryMock.getNewPatchFinder()).thenReturn(patchFinderMock);
         when(purlMapperMock.getPurlBase(Mockito.any(Vulnerability.class))).thenReturn("pkg:pypi/django");
 
-        pm.getVulnerabilitiesFromParsers(false);
+        pm.getVulnerabilitiesFromParsers(false, 1);
 
         // EXPECTED
         var vE1 = new Vulnerability("CVE-TEST-1");
@@ -195,7 +195,7 @@ public class ParserManagerTest {
         when(purlMapperMock.getPurlBase(Mockito.any(Vulnerability.class))).thenReturn("pkg:pypi/django");
         when(patchFinderFactoryMock.getNewPatchFinder()).thenReturn(patchFinderMock);
 
-        pm.getVulnerabilitiesFromParsers(true);
+        pm.getVulnerabilitiesFromParsers(true, 1);
 
         // EXPECTED
         var vE1 = new Vulnerability("CVE-TEST-1");

